### DISCUSSION TITLE
Using `ml_dtypes.bfloat16` with `nb::ndarray`

### DISCUSSION THREAD

**Discussion by awni** [wjakob/nanobind#560] [Q&A]
Using `bfloat16` from `ml_dtypes` as the type of a `np.ndarray` is not cast to a `nd::array`. 

Is there a way to use the [non-standard types](https://nanobind.readthedocs.io/en/latest/ndarray.html#nonstandard-arithmetic-types) to get this to work? 

For example, given this extension:

```c++
#include <nanobind/nanobind.h>                                                 
#include <nanobind/ndarray.h>                                                  
                                                                               
namespace nb = nanobind;                                                       
                                                                               
void check(nb::ndarray<nb::ro, nb::c_contig, nb::device::cpu>) {               
  std::cout << "Got ND Array? " << std::endl;                                  
}                                                                              
                                                                               
NB_MODULE(ext, m) {                                                            
    m.def("check", &check);                                                    
}       
```

The following code throws an error on the `bfloat` array (which is expected).

```python
import ext
from ml_dtypes import bfloat16
import numpy as np

x = np.array(1., dtype=np.float16)
ext.check(x)

x = np.array(1., dtype=bfloat16)
# Error:
# TypeError: check(): incompatible function arguments. The following argument types are supported:
#    1. check(arg: typing.Union[float, ndarray[writable=False, order='C', device='cpu']], /) -> None
#
# Invoked with types: ndarray
ext.check(x)
```

**Comment by wjakob** [comment]
What would be the C++ type associated with `bfloat`? If you have that, then it is likely possible to use the documented approach.

**Reply by awni** [reply]
That's what I'm looking for now. We have a bfloat16  in MLX already that is registered as a custom type with `nb::ndarray`: https://github.com/ml-explore/mlx/blob/main/python/src/convert.cpp#L19-L26

I am trying to figure out how we can map the `ml_dtype.bfloat16` to the MLX bfloat16 type.

**Reply by awni** [reply]
That's the ml_dtype bfloat, it looks like it's using Eigen: https://github.com/jax-ml/ml_dtypes/blob/main/ml_dtypes/_src/dtypes.cc#L40-L60

**Reply by awni** [reply]
I'm not sure I follow how the documented approach will work here. It looks like nanobind doesn't cast the incoming `np.ndarray` to a `nb::ndarray` (but it casts fine for fp16 which is also non-standard). If I register the `Eigen::bfloat16` with Nanobind would that cause the cast to succeed? 

```c++
namespace nanobind {
    template <> struct ndarray_traits<Eigen::bfloat16> {
        static constexpr bool is_complex = false;
        static constexpr bool is_float   = true;
        static constexpr bool is_bool    = false;
        static constexpr bool is_int     = false;
        static constexpr bool is_signed  = true;
    };
}
```

**Reply by makslevental** [reply]
FYI @awni's idea does not work i.e., I have the exact same issue and tried it out (after building/linking Eigen) and I get basically no difference in the error reported:
```cpp
"create_elements_attr", [](MlirLocation loc, MlirType t, const nb::ndarray<> &bytes)
```
produces
```
E           TypeError: create_elements_attr(): incompatible function arguments. The following argument types are supported:
E               1. create_elements_attr(arg0: MlirLocation, arg1: MlirType, arg2: ndarray[], /) -> MlirAttribute
E           
E           Invoked with types: odie.mlir._mlir_libs._mlir.ir.Location, odie.mlir._mlir_libs._mlir.ir.RankedTensorType, ndarray
```
and
```cpp
"create_elements_attr", [](MlirLocation loc, MlirType t, const nb::ndarray<nb::ro, nb::device::cpu, nb::c_contig> &bytes)
```
produces
```
E           TypeError: create_elements_attr(): incompatible function arguments. The following argument types are supported:
E               1. create_elements_attr(arg0: MlirLocation, arg1: MlirType, arg2: ndarray[order='C', device='cpu', writable=False], /) -> MlirAttribute
E           
E           Invoked with types: odie.mlir._mlir_libs._mlir.ir.Location, odie.mlir._mlir_libs._mlir.ir.RankedTensorType, ndarray
```

I'm guessing this is because for `ndarray_traits<Eigen::bfloat16>` needs to be specialized in the same place where `struct TypeDescriptor<bfloat16> : CustomFloatType<bfloat16>` is specialized (i.e., `ml_dtypes`).

**Comment by hpkfft** [comment]
The Python buffer protocol does not support bfloat16.  There's no format character for it.
See: https://docs.python.org/3/library/struct.html#format-characters

DLPack does support bfloat16, and it has a dtype code for it, but Numpy does not support it:
```
>>> import numpy as np
>>> import ml_dtypes
>>> a = np.array([3, 7], dtype=ml_dtypes.bfloat16)
>>> a
array([3, 7], dtype=bfloat16)
>>> a.__dlpack__()
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    a.__dlpack__()
    ~~~~~~~~~~~~^^
BufferError: DLPack only supports signed/unsigned integers, float and complex dtypes.
```
As far as I can tell, if NumPy were to support exporting the array as a DLPack capsule, nanobind would accept it.

If you don't need arrays of uint16, then you can cheat by adopting your own convention that arrays of uint16 are understood to represent bfloat16 values:
```
namespace nanobind::detail {
    template<> struct dtype_traits<_Float16> {
        static constexpr dlpack::dtype value {
            (uint8_t) dlpack::dtype_code::Float, // type code
            16, // size in bits
            1   // lanes (simd), usually set to 1
        };
        static constexpr auto name = const_name("float16");
    };
    template<> struct dtype_traits<__bf16> {
        static constexpr dlpack::dtype value {
            (uint8_t) dlpack::dtype_code::Bfloat, // type code
            16, // size in bits
            1   // lanes (simd), usually set to 1
        };
        static constexpr auto name = const_name("bfloat16");
    };
}

NB_MODULE(my_extension, m) {
    m.def("inspect", [](const nb::ndarray<nb::ro, nb::device::cpu>& a) {
        printf("Array data pointer : %p\n", a.data());
        printf("Array dimension : %zu\n", a.ndim());
        for (size_t i = 0; i < a.ndim(); ++i) {
            printf("Array dimension [%zu] : %zu\n", i, a.shape(i));
            printf("Array stride    [%zu] : %zd\n", i, a.stride(i));
        }
        if (a.dtype() == nb::dtype<double>()) {
            const double* array = static_cast<const double*>(a.data());
            printf("First element is float64: %f\n", array[0]);
        } else if (a.dtype() == nb::dtype<float>()) {
            const float* array = static_cast<const float*>(a.data());
            printf("First element is float32: %f\n", array[0]);
        } else if (a.dtype() == nb::dtype<_Float16>()) {
            const _Float16* array = static_cast<const _Float16*>(a.data());
            double value = static_cast<double>(array[0]);
            printf("First element is float16: %f\n", value);
        } else if (a.dtype() == nb::dtype<__bf16>()) {
            const __bf16* array = static_cast<const __bf16*>(a.data());
            double value = static_cast<double>(array[0]);
            printf("First element is bfloat16: %f\n", value);
        } else if (a.dtype() == nb::dtype<uint16_t>()) {
            const __bf16* array = static_cast<const __bf16*>(a.data());
            double value = static_cast<double>(array[0]);
            printf("First element is uint16, but as bfloat16: %f\n", value);
        }
    });
}
```
which can be used as follows:
```
>>> import numpy as np
>>> import ml_dtypes
>>> import my_extension as me
>>> a = np.array([3, 7], dtype=np.float64)
>>> me.inspect(a)
Array data pointer : 0x55f70dfa9d50
Array dimension : 1
Array dimension [0] : 2
Array stride    [0] : 1
First element is float64: 3.000000
>>> a = np.array([3, 7], dtype=np.float32)
>>> me.inspect(a)
Array data pointer : 0x55f70de13150
Array dimension : 1
Array dimension [0] : 2
Array stride    [0] : 1
First element is float32: 3.000000
>>> a = np.array([3, 7], dtype=np.float16)
>>> me.inspect(a)
Array data pointer : 0x55f70dfbe5b0
Array dimension : 1
Array dimension [0] : 2
Array stride    [0] : 1
First element is float16: 3.000000
>>> a = np.array([3, 7], dtype=ml_dtypes.bfloat16)
>>> me.inspect(a)
Traceback (most recent call last):
  File "<python-input-10>", line 1, in <module>
    me.inspect(a)
    ~~~~~~~~~~^^^
TypeError: inspect(): incompatible function arguments. The following argument types are supported:
    1. inspect(arg: ndarray[device='cpu', writable=False], /) -> None

Invoked with types: ndarray
>>> me.inspect(a.view(np.uint16))
Array data pointer : 0x55f70e03afe0
Array dimension : 1
Array dimension [0] : 2
Array stride    [0] : 1
First element is uint16, but as bfloat16: 3.000000
```

**Comment by hpkfft** [comment]
If you don't like calling `a.view(np.uint16)` in Python, you can call it using nanobind:
```
void test(nb::object a) {
    bool is_bfloat16 = false;
    try {
        nb::object ml_dtypes = nb::module_::import_("ml_dtypes");
        nb::object bfloat16 = ml_dtypes.attr("bfloat16");
        nb::object dtype = a.attr("dtype");
        is_bfloat16 = dtype.equal(bfloat16);
    } catch (const nb::python_error &e) {
        is_bfloat16 = false;
    }
    if (!is_bfloat16) throw nb::next_overload();

    nb::object u = a.attr("view")("uint16");
    auto ndarray = nb::cast<nb::ndarray<const uint16_t, nb::device::cpu>>(u);
    const uint16_t* data = ndarray.data();
    std::size_t size = ndarray.size();
    printf("array: ");
    for (std::size_t i = 0; i < size; ++i) {
        __bf16 value = std::bit_cast<__bf16>(data[i]);
        printf(" %f", static_cast<double>(value));
    }
    printf("\n");
}

NB_MODULE(my_extension, m) {
    m.def("test", test);
}
```
which seems to work:
```
>>> import numpy as np
>>> import ml_dtypes
>>> import my_extension as me
>>> a = np.array([3, 7], dtype=ml_dtypes.bfloat16)
>>> me.test(a)
array:  3.000000 7.000000
```

I didn't think too much about the error checking, but hopefully you get the idea.

**Comment by makslevental** [comment]
@hpkfft 

> The Python buffer protocol does not support bfloat16. There's no format character for it.
>See: https://docs.python.org/3/library/struct.html#format-characters
> DLPack does support bfloat16, and it has a dtype code for it, but Numpy does not support it:

Sure but that's the whole point of `ml_dtypes` right? They register descriptors for all the "ml dtypes". So "Numpy does not support it" is "ill-posed" because Numpy in effect does support it via external descriptors (which the `ml_dtypes` folks have been so kind as to supply). 

The issue here in nanobind is that *nanobind doesn't actually support dlpack*: 

https://github.com/wjakob/nanobind/blob/66dfe77f42b76590076674e21648da6b1c67979a/include/nanobind/ndarray.h#L10-L11

So this is a case of "we have dlpack at home" but dlpack at home doesn't support interop with `ml_dtypes`. Moreover, it's not clear to me that the bespoke "dlpack" here is actually mutually compatible with actual [DLPack](https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h), which presumably *would* actually seamlessly work with [Numpy](https://numpy.org/doc/2.0/reference/generated/numpy.from_dlpack.html). Not sure the reason for the design choice of rolling your own dlpack 🤷. 

Anyway I resolved my issue (the reason I bumped this old issue) by casting to `PyArrayObject`

```c++
  auto validateAndCastNumpyArray = [](PyObject *obj) {
    if (!PyArray_Check(obj))
      throw std::invalid_argument("expected numpy array");
    const PyArrayObject *arr = reinterpret_cast<PyArrayObject *>(obj);
    if (!PyArray_IS_C_CONTIGUOUS(arr))
      throw std::invalid_argument("expected contiguous array");
    if (!PyArray_ISBEHAVED_RO(arr))
      throw std::invalid_argument("expected read-only array");
    return arr;
  };
```  

which was good enough because I just needed the bytes anyway, not the bfloat16 value or any some such.

**Reply by hpkfft** [reply]
No, you're wrong.  The following does not involve nanobind in any way whatsoever:
```
>>> import numpy as np
>>> import ml_dtypes
>>> a = np.array([3, 7], dtype=ml_dtypes.bfloat16)
>>> a
array([3, 7], dtype=bfloat16)
>>> a.__dlpack__()
Traceback (most recent call last):
  File "<python-input-4>", line 1, in <module>
    a.__dlpack__()
    ~~~~~~~~~~~~^^
BufferError: DLPack only supports signed/unsigned integers, float and complex dtypes.
```

**Reply by wjakob** [reply]
> Moreover, it's not clear to me that the bespoke "dlpack" here is actually mutually compatible with actual [DLPack](https://github.com/dmlc/dlpack/blob/main/include/dlpack/dlpack.h), which presumably would actually seamlessly work with [Numpy](https://numpy.org/doc/2.0/reference/generated/numpy.from_dlpack.html).

The DLPack ABI relevant to nanobind boils down to 2 C-style structs, and a few enum entries (which we have copied instead of adding another submodule dependency that seems infinitely worse). There isn't that much here that can go wrong. 🤷

**Reply by makslevental** [reply]
Fair enough - I wrongly assumed `__dlpack__` would work but I guess it boils down to [numpy not recognizing the `type_num` associated with bfloat16](https://github.com/numpy/numpy/blob/980c1088b63765e915065c34b9584a2483bbb2a5/numpy/_core/include/numpy/ndarraytypes.h#L1666) (presumably set somewhere in `ml_dtypes`). 

Question for y'all: if `ml_dtypes` itself adopts `nanobind` and adds `ndarray_traits<Eigen::bfloat16>` will that work in other extension modules which use `nb::ndarray` and expect `bfloat16` dtype arrays? Or is this one of those things where (because of `typeid` or something?) you'd have to compile all of the above (`ml_dtypes` and extension module `foo` and `ndarray_traits<Eigen::bfloat16>`) into one single extension module?

**Reply by wjakob** [reply]
So nanobind actually never generates a `dtype` (Python) object. It has no concept of that, and there is no builtin bindings for such a thing in the nanobind ndarray support. The only thing that nanbind knows about dtypes is based on what's encoded in a C-style structure, either one represented by the buffer protocol `struct`, or one represented by the DLPack data type `struct`. So I don't think `ml_dtype` switching to nanobind would help accomplish anything. The best long term solution is if the buffer protocol was updated with new characters to represent these new and emerging numerical types.

**Reply by wjakob** [reply]
One more detail: `ml_dtypes` is definitely a kind of workaround. It uses `PyArray_RegisterDataType` to register a custom data type with arithmetic type `V` (void). So NumPy actually really doesn't understand the number type it is operating on. The only reason it works is because the library also implements handlers for all standard operations. Obviously this is not a good starting point for inter-op with other libraries that actually do understand these types.

**Reply by wjakob** [reply]
In addition, this library does not work for NumPy2 at the moment. So things are kind of a mess. I don't think it makes sense for nanobind to do anything special about this library.

**Reply by hpkfft** [reply]
I agree.  I'd like to note that Intel and AMD are adding AVX10 to x86_64, and AVX10.2 adds arithmetic instructions (add, sub, mul, FMA, div, sqrt) for bfloat16.  Also, AVX10.2 has 512-bit SIMD, and it's for both server CPUs (Diamond Rapids) and client CPUs (Nova Lake).  For AArch64, the bfloat16 instructions are added in the feature FEAT_SVE_B16B16.
Given future CPU hardware support, it may be reasonable to predict that the C standard will endorse a type for bfloat16, Python will add a format character for it, and NumPy itself will add a dtype for it.

BTW, the "inspect" code I shared above works with PyTorch:
```
>>> import torch
>>> import my_extension as me
>>> a = torch.tensor([3.25, 7.5], dtype=torch.bfloat16)
>>> a
tensor([3.2500, 7.5000], dtype=torch.bfloat16)
>>> me.inspect(a)
Array data pointer : 0x55db0ad95400
Array dimension : 1
Array dimension [0] : 2
Array stride    [0] : 1
First element is bfloat16: 3.250000
```
and JAX:
```
>>> import jax
>>> from jax import numpy as jnp
>>> import my_extension as me
>>> a = jnp.array([3.2500, 7.5000], dtype=jnp.bfloat16)
>>> a
Array([3.25, 7.5], dtype=bfloat16)
>>> me.inspect(a)
Array data pointer : 0x559bb15402c0
Array dimension : 1
Array dimension [0] : 2
Array stride    [0] : 1
First element is bfloat16: 3.250000
```
without having to use `uint16`.
